File: Trabajo_6/CPP/Actividades/Actividad_2.cpp
#include <iostream>

using namespace std;

class cliente{

protected:
    string nombre;
    int telefono{};
    string correo;
    string cuentaBancaria;
    double monto{};

    static string encriptarInformacion(string informacion, int desplazamiento=3) {
        string resultado;
        for (int i = 0; i < informacion.length(); i++) {
            if (isdigit(informacion[i]))
                resultado += char(int(informacion[i] + desplazamiento - 48) % 10 + 48);
            else if (isupper(informacion[i]))
                resultado += char(int(informacion[i] + desplazamiento - 65) % 26 + 65);
            else
                resultado += char(int(informacion[i] + desplazamiento - 97) % 26 + 97);
        }
        return resultado;
    }

public:

    cliente(string nombre, int telefono, const string& correo, const string& cuentaBancaria, double monto){
        this -> nombre = nombre;
        this -> telefono = telefono;
        this -> correo = correo;
        this -> cuentaBancaria = cuentaBancaria;
        this -> monto = monto;
    }


    friend ostream &operator<<(ostream &os, const cliente &cliente) {
        os << "Nombre: " << cliente.nombre << "\n"
           << "Telefono: " << cliente.telefono << "\n"
           << "Correo: " << cliente.correo << "\n"
           << "Cuenta Bancaria: " << cliente.cuentaBancaria << "\n"
           << "Monto: " << cliente.monto;
        return os;
    }
};



class clienteSeguro : public cliente {
private:
    string claveEncriptada;

    static string descriptarInformacion(string informacion, int desplazamiento=3) {
        string resultado;
        for (int i = 0; i < informacion.length(); i++) {
            if (isdigit(informacion[i]))        //verificar si es un digito
                resultado += char(int(informacion[i] - desplazamiento - 48 + 10) % 10 + 48);
            else if (isupper(informacion[i]))   // verificar si es una letra mayuscula
                resultado += char(int(informacion[i] - desplazamiento - 65 + 26) % 26 + 65);
            else                                   // verificar si es una letra minuscula
                resultado += char(int(informacion[i] - desplazamiento - 97 + 26) % 26 + 97);
        }
        return resultado;
    }


public:
    clienteSeguro(string nombre, int telefono, string correo, string cuentaBancaria, double monto, string clave)
            : cliente(encriptarInformacion(nombre), telefono, encriptarInformacion(correo), encriptarInformacion(cuentaBancaria), monto){

        this -> claveEncriptada = encriptarInformacion(std::move(clave));

    }

    string getClaveEncriptada()
    {
        return claveEncriptada;
    }

    static string encriptarInformacion(string informacion, int desplazamiento=3) {
        string resultado;
        for (int i = 0; i < informacion.length(); i++) {
            if (isdigit(informacion[i]))
                resultado += char(int(informacion[i] + desplazamiento - 48) % 10 + 48);
            else if (isupper(informacion[i]))
                resultado += char(int(informacion[i] + desplazamiento - 65) % 26 + 65);
            else
                resultado += char(int(informacion[i] + desplazamiento - 97) % 26 + 97);
        }
        return resultado;
    }

    bool verificarAutenticidad(const string& claveIngresada) {
        return claveIngresada == descriptarInformacion(claveEncriptada);
    }
};



int main(){

    clienteSeguro grossman = clienteSeguro("Juan Perez",
                                           987654321,
                                           "Juan.Perez@gmail.com",
                                           "123456789",
                                           1000.0,
                                           "1234");


    cout << grossman << endl;

    cout << "Clave encriptada: " << grossman.getClaveEncriptada() << endl;

    cout << "Ingrese la clave para verificar la autenticidad:" << endl;
    string claveIngresada;
    cin >> claveIngresada;

    if (grossman.verificarAutenticidad(claveIngresada)) {
        cout << "La clave ingresada es correcta" << endl;
    } else {
        cout << "La clave ingresada es incorrecta" << endl;
    }
}

File: Trabajo_7/JAVA/Actividades/Actividad_2.java
/*

package Trabajo_7.Actividades;


import java.util.Scanner;

class Cliente {
    protected String nombre;
    protected int telefono;
    protected String correo;
    protected String cuentaBancaria;
    protected double monto;

    Cliente(String nombre, int telefono, String correo, String cuentaBancaria, double monto) {
        this.nombre = nombre;
        this.telefono = telefono;
        this.correo = correo;
        this.cuentaBancaria = cuentaBancaria;
        this.monto = monto;
    }

    @Override
    public String toString() {
        return "Nombre: " + nombre + "\n" +
                "Telefono: " + telefono + "\n" +
                "Correo: " + correo + "\n" +
                "Cuenta Bancaria: " + cuentaBancaria + "\n" +
                "Monto: " + monto;
    }
}

class ClienteSeguro extends Cliente {
    private final String claveEncriptada;

    ClienteSeguro(String nombre, int telefono, String correo, String cuentaBancaria, double monto, String clave) {
        super(nombre, telefono, correo, cuentaBancaria, monto);
        this.nombre = encriptarInformacion(nombre,3);
        this.telefono = telefono;
        this.correo = encriptarInformacion(correo,3);
        this.cuentaBancaria = encriptarInformacion(cuentaBancaria,3);
        this.monto = monto;
        this.claveEncriptada = encriptarInformacion(clave,3);
    }

    String getClaveEncriptada() {
        return claveEncriptada;
    }

    static String encriptarInformacion(String informacion, int desplazamiento) {
        StringBuilder resultado = new StringBuilder();
        for (int i = 0; i < informacion.length(); i++) {
            if (Character.isDigit(informacion.charAt(i))) // verificar si es un digito
                resultado.append((char) ((informacion.charAt(i) + desplazamiento - '0') % 10 + '0'));
            else if (Character.isUpperCase(informacion.charAt(i))) // verificar si es una letra mayuscula
                resultado.append((char) ((informacion.charAt(i) + desplazamiento - 'A') % 26 + 'A'));
            else  // verificar si es una letra minuscula
                resultado.append((char) ((informacion.charAt(i) + desplazamiento - 'a') % 26 + 'a'));
        }
        return resultado.toString();
    }

    private static String descriptarInformacion(String informacion, int desplazamiento) {
        StringBuilder resultado = new StringBuilder();
        for (int i = 0; i < informacion.length(); i++) {
            if (Character.isDigit(informacion.charAt(i))) // verificar si es un digito
                resultado.append((char) ((informacion.charAt(i) - desplazamiento - 48 + 10) % 10 + 48));
            else if (Character.isUpperCase(informacion.charAt(i))) // verificar si es una letra mayuscula
                resultado.append((char) ((informacion.charAt(i) - desplazamiento - 65 + 26) % 26 + 65));
            else // verificar si es una letra minuscula
                resultado.append((char) ((informacion.charAt(i) - desplazamiento - 97 + 26) % 26 + 97));
        }
        return resultado.toString();
    }

    boolean verificarAutenticidad(String claveIngresada) {
        return claveIngresada.equals(descriptarInformacion(claveEncriptada,3));
    }
}

public class Actividad_1 {
    public static void main(String[] args) {
        ClienteSeguro grossman = new ClienteSeguro("Juan Perez",
                987654321,
                "Juan.Perez@gmail.com",
                "123456789",
                1000.0,
                "1234");

        System.out.println(grossman);

        System.out.println("Clave encriptada: " + grossman.getClaveEncriptada());

        System.out.println("Ingrese la clave para verificar la autenticidad:");
        Scanner scanner = new Scanner(System.in);
        String claveIngresada = scanner.nextLine();

        if (grossman.verificarAutenticidad(claveIngresada)) {
            System.out.println("La clave ingresada es correcta");
        } else {
            System.out.println("La clave ingresada es incorrecta");
        }
    }
}
*/
